java-squid/src/main/java/org/sonar/java/cfg/LiveVariables.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * sonarqube@googlegroups.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.cfg;
+
+import com.google.common.collect.Sets;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import java.io.PrintStream;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+
+public class LiveVariables {
+
+  private final CFG cfg;
+  private final Map<CFG.Block, Set<Symbol>> in = new HashMap<>();
+  private final Map<CFG.Block, Set<Symbol>> out = new HashMap<>();
+
+  private LiveVariables(CFG cfg) {
+    this.cfg = cfg;
+  }
+
+  public Set<Symbol> getOut(CFG.Block block) {
+    return out.get(block);
+  }
+
+  public static LiveVariables analyze(CFG cfg) {
+    LiveVariables liveVariables = new LiveVariables(cfg);
+
+    Map<CFG.Block, Set<Symbol>> kill = new HashMap<>();
+    Map<CFG.Block, Set<Symbol>> gen = new HashMap<>();
+    for (CFG.Block block : cfg.blocks) {
+      Set<Symbol> blockKill = new HashSet<>();
+      Set<Symbol> blockGen = new HashSet<>();
+
+      // process elements from bottom to top
+      Set<Tree> assignmentLHS = new HashSet<>();
+      for (Tree element : block.elements) {
+        Symbol symbol;
+        switch (element.kind()) {
+          case ASSIGNMENT:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 13 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)